Made a first (partial) attempt at implementing touchbar hold behaviour

Joel-Rogers opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->

<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description

<!--- Describe your changes in detail here. -->
touch_encoder.c: Added two new booleans which are toggled on/off at start/end
    of swiping/holding detection. Swiping is given priority over holding for any
    continuous touches, so moving out of the deadzone at any point during a single
    touch should lock the user into the swiping code until they stop touching.
    I've tried to keep my additions as separate from existing as possible for now
    to keep any interference to a minimum and tracking easy, but there are a few
    places where it could be refactored/streamlined. I've added a couple of new
    functions which closely mirror the existing tapping ones but press and release
    explicitly so I could logically handle long, static touches as well as
    transition-to-swipes - it's also required since obviously hold behaviour
    should start ASAP, and certainly before touch-release.
    I've made no attempt to deal with any slave behaviour, and
    touch_slave_state.taps in particular is in a few places it probably
    shouldn't be - I don't understand it well enough to know if it can be
    hijacked/overloaded with the hold tracking, or if this needs its own variable
    and/or transport.
    I'm not sure how reliably touch_encoder_update_position() will be called if
    the user is extremely still while holding - needs real-world testing perhaps.
    It only needs calling once anyway for holding.
    touch_encoder.h: Added declarations for the new weak functions, which are
    analogous to the tap ones.
    rev1.c: Added two new process_encoder_matrix functions which press and release
    keys separately (if I understand action_exec correctly), and the default
    weak function definitions which just call them as for taps. Closer
    integration is possible here if/when the new code is working.
    rules.mk: Added TOUCHBAR_HOLD_ENABLE as a toggle for the holding behaviour -
    the booleans will still be toggled a bit, but no functions should get called.
    keymaps/Joel-Rogers/keymap.c: My keymap, with momentary layers on the
    touchbar - should work once the hold behaviour does.

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ ] Bugfix
- [ x] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [ ] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

XScorpion2 committed:
RGBKB Mun Keyboard

XScorpion2 committed:
Latest qmk changes

XScorpion2 committed:
Split RGB Matrix

XScorpion2 committed:
RGBKB Mun Keyboard

Enabling encoders by default

Updates to Mun

Keymap, config, and RGB layout updates.

Mun Keymap & Encoder updates

Legonut committed:
Updates to Mun

Keymap, config, and RGB layout updates.

XScorpion2 committed:
Suspend State sync for rgb matrix

noroadsleft committed:
Branch point for 2021 May 29 Breaking Change

zvecr committed:
Overhaul bootmagic logic to have single entrypoint (#8532)

* Relocate bootmagic logic to have single entrypoint

* Align init of layer state

zvecr committed:
ARM - Refactor SLEEP_LED to support more platforms (#8403)

* Initial refactor of ARM SLEEP_LED to enable more platforms

* fix build issues

* Disable SLEEP_LED for boards with no caps lock code

* Enable GPT14 for boards with caps lock code and SLEEP_LED enabled

* Enable GPT for boards with caps lock code and SLEEP_LED enabled

zvecr committed:
Refactor of USB code within split_common (#11890)

* Initial refactor of usb code within split_common

* Add headers

* Correct disable condition

* Format

* Align func name

github-actions[bot] committed:
Format code according to conventions (#12056)

Co-authored-by: QMK Bot <hello@qmk.fm>

XScorpion2 committed:
Enabling encoders by default

Legonut committed:
Initial files for Sol 3

To be wrapped into original Sol folder

Legonut committed:
Update keymap, enable audio

Legonut committed:
Fix default layers

Legonut committed:
Keymap updates

kageurufu committed:
Sol3 matrix encoders support
Mun encoder functional

kageurufu committed:
Restore correct default mun keymap

Legonut committed:
Keymap updates

Legonut committed:
Keymap updates

Legonut committed:
keymap updates

XScorpion2 committed:
Split support for touch encoders

XScorpion2 committed:
hoisted encoder functionality up to keyboard level

Legonut committed:
Create temporary Zygomorth 2 folder

XScorpion2 committed:
Merge branch 'rgbkb/mun' into sol3

XScorpion2 committed:
Updated Sol3 and Zygomorph2 with changes from Mun

XScorpion2 committed:
Fixing Lego's spelling

XScorpion2 committed:
Added touch toggling
Removed oled debug information
Cleaned up keymap functions

Legonut committed:
Update keymap to reflect sol3 changes

Legonut committed:
Remove saturation from touchbars

Legonut committed:
Change encoders to media controls

Legonut committed:
Update touchbar controls

kageurufu committed:
RGBKB common module

XScorpion2 committed:
Fix touch disable & slave slider spam

Legonut committed:
Update keymap, remove touch buttons for now

XScorpion2 committed:
Added brightness limit

Legonut committed:
Disable left hand encoders for now, rassign pins

kageurufu committed:
Fix the encoder matrix reset to pull-up instead of down

kageurufu committed:
Add kageurufu sol3 keymap

XScorpion2 committed:
Mun code fixes & xulkal keymap

Legonut committed:
Fix layout, disable encoders fow now

kageurufu committed:
Add Sol3 LED TURBO Dip switch support

This adds the rgb_matrix LED_FLAG_LIMIT_BRIGHTNESS flag,
 and the RGB_MATRIX_LIMIT_BRIGHTNESS define to set a
 maximum brightness while limited.

kageurufu committed:
Revert to sym_defer_pk for now

XScorpion2 committed:
Fixed define, decreased touch sensitivity, fixed touch freak out

XScorpion2 committed:
fixing my slave oled for now

Legonut committed:
Fix mun keymaps

XScorpion2 committed:
host touch bar status on oled

XScorpion2 committed:
Fix right touch encoder tap index

XScorpion2 committed:
Hacked QMK to allow tap of QMK keycodes

XScorpion2 committed:
Merge branch 'develop' into development

XScorpion2 committed:
Fixing transport's led/rgb matrix suspend state logic

XScorpion2 committed:
More tap code support and encoder sync init

XScorpion2 committed:
Debounce improvements, encoder handling improvements, default oled improvements

XScorpion2 committed:
Minor code cleanup

kageurufu committed:
Sol3 encoder matrix support.

This also brings Via support to a minimum level

Include @XScorpion2's new OLED Menu
Enable the thumb encoder grounded buttons, although
 it doesnt work right on the slave half yet

XScorpion2 committed:
qmk configurator files, keymap.json to move into configurator git

XScorpion2 committed:
File rename to match qmk configurator naming conventions

Joel-Rogers committed:
Pre-touchbar fix personal tweaks

Joel-Rogers committed:
Add keymap

Joel-Rogers committed:
Merge and resolve touch sensitivity conflict

Joel-Rogers committed:
colejdr v4 keymap commit

Joel-Rogers committed:
Make QOL changes for toggled layers

Joel-Rogers committed:
First attempt at touchbar hold behaviour

touch_encoder.c: Added two new booleans which are toggled on/off at start/end
of swiping/holding detection. Swiping is given priority over holding for any
continuous touches, so moving out of the deadzone at any point during a single
touch should lock the user into the swiping code until they stop touching.
I've tried to keep my additions as separate from existing as possible for now
to keep any interference to a minimum and tracking easy, but there are a few
places where it could be refactored/streamlined. I've added a couple of new
functions which closely mirror the existing tapping ones but press and release
explicitly so I could logically handle long, static touches as well as
transition-to-swipes - it's also required since obviously hold behaviour
should start ASAP, and certainly before touch-release.
I've made no attempt to deal with any slave behaviour, and
touch_slave_state.taps in particular is in a few places it probably
shouldn't be - I don't understand it well enough to know if it can be
hijacked/overloaded with the hold tracking, or if this needs its own variable
and/or transport.
I'm not sure how reliably touch_encoder_update_position() will be called if
the user is extremely still while holding - needs real-world testing perhaps.
It only needs calling once anyway for holding.
touch_encoder.h: Added declarations for the new weak functions, which are
analogous to the tap ones.
rev1.c: Added two new process_encoder_matrix functions which press and release
keys separately (if I understand action_exec correctly), and the default
weak function definitions which just call them as for taps. Closer
integration is possible here if/when the new code is working.
rules.mk: Added TOUCHBAR_HOLD_ENABLE as a toggle for the holding behaviour -
the booleans will still be toggled a bit, but no functions should get called.
keymaps/Joel-Rogers/keymap.c: My keymap, with momentary layers on the
touchbar - should work once the hold behaviour does.

Joel-Rogers committed:
Fix small errors and verify compilation success

Joel-Rogers commented:
At the moment, release is handled well, but initiating holding is difficult. Not sure if this is a deadzone or logic issue

XScorpion2 commented on keyboards/rgbkb/common/touch_encoder.c:
```diff
@@ -114,10 +114,18 @@ enum {  // QT2120 registers
 // touch_raw & touch_processed store the Detection Status, Key Status (x2), and Slider Position values
 uint8_t  touch_raw[4]       = { 0 };
 uint8_t  touch_processed[4] = { 0 };
+// JoelR note - I have interpreted touch_raw[0] as a boolean of whether or not the touchbar is currently being touched - if that assumption is wrong, then the logic I put in for holding will be wrong
```
This is an incorrect assumption. touch_raw[0] is the detection status of the chip which is 8 bits long (1 byte) and contains multiple flags and you have to test each flag individually: CALIBRATE, OVERFLOW, SDET, and TDET
http://ww1.microchip.com/downloads/en/DeviceDoc/doc9634.pdf

Though I think most of these comments are out of date looking at the recent code.

Joel-Rogers replied:
Ah, I see - it might have been more accurate to say that I interpreted "touch_raw[0] & SLIDER_BIT" as the above boolean. Which is perhaps the case? I never actually inserted it anywhere, I just hijacked the existing slider function, but this was how I rationalised that decision.

Joel-Rogers replied:
I think there will always be a sensitivity trade-off between holding and swiping, but could maybe be offset a bit by adding a second TOUCH_TERM-style variable?

XScorpion2 reviewed (commented):
I see roughly what you are wanting to implement, need a bit more time to think through this logic a bit more and ensure there are no side effects. (IE: firing hold when you intended to swipe because it fires too quickly, etc)

Joel-Rogers committed:
Update keymap w home row mods and more layers

Added momentary layer control to main keys to remove dependency on touchbar

Joel-Rogers committed:
Keymap tweaks; fix tap-dance toggle

Joel-Rogers committed:
Touchbar fix errors, sensitivity; tidy comments

Fixed a missing is_holding boolean reset, and removed a (seemingly) unnecessary
logical check in touch_encoder_update(). These changes considerably improve
responsiveness to/reliability of holding. Removed comments leftover from
early testing to improve legibility.
